### DISCUSSION TITLE
GSoC 2026: Integrating OLMo 7B Model into DeepChem

### DISCUSSION THREAD

**Discussion by priyadp1** [deepchem/deepchem#5114] [Show and tell]
### **Introduction**
My name is Prisha Priyadarshini and I'm a senior Computer Science major and Mathematics minor at Rutgers University-- New Brunswick and this summer I had the opportunity to work alongside DeepChem members to integrate the OLMo 7B model into DeepChem which is the first time a large model is being integrated into the DeepChem framework. Before I started working on this project DeepChem had support for transformer-based models through its HuggingFaceModel wrapper. However, support for modern LLMs was still limited. To make sure that DeepChem had support for modern LLMs such as OLMo 7B, I integrated a generate function and added task=causal_lm in hf_models.py, built a OLMo wrapper class using custom linear heads to support single task regression, single task classification, multitask regression, multitask classification, and causal lm support, and then I did benchmarking across 9 datasets.

### **PR #4992  (Generate Function , Merged)**
In this now merged PR, I added the generate function and task = causal_lm into hf_models.py. I added the causal_lm task into the prepare_batch() function alongside the already supported task types in DeepChem (mtr and mlm). My code that integrates causal_lm into hf_models.py takes tokenized text, uses the tokens as next-token prediction targets, ignores padding in the loss, and packages everything for causal_lm training. In my generate() function, I used my updated prepare_batch() function to help DeepChem's hf_models.py generate text using HuggingFace's text generation pipeline. My generate() function uses End of Sequence (EOS) padding because causal_lm models such as GPT-2 don't have a designated padding token. I also added left padding because for decoder-only models such as OLMo, it ensures generation starts after the actual end of each prompt. I then had to add MPS-specific device handling since HuggingFace's generate method does not currently support MPS versions less than 14. So, if the MPS version is less than 14, I move the tensors to CPU. After device detection, I finally do the actual generation code by splitting the Dataset into batches and calling HuggingFace's text generation system. After generation is done, I restore the model/tokenizer to their state before generate() was called.

### **PR #5093 (Fixing HF Tests , Merged)**
In collaboration with @sudo-rsingh , @Zrahay , and @shubham-bari I created this now-merged PR based off of @sudo-rsingh's original fixes for the HuggingFace tests. Originally, @sudo-rsingh implemented a fix that pinned an older commit from ibm/MoLFormer-XL-both-10pct. This was necessary because later versions of the remote code require transformers>=4.53 which isn't supported by DeepChem. His fix fixed HF-tests 3.8 and 3.10 I then took his code and fixed the formatting which fixed HF-Test 3.11. However, HF-Test 3.9 which is windows-based was still not working due to an SSL error. To fix this, I patched the SSL configuration to use certifi’s trusted CA certificate bundle for HTTPS verification. This resolves certificate verification errors while preserving secure SSL validation. Now, all the HF-tests work in the DeepChem CI.

### **PR #5001 (OLMo Wrapper Part 1 , Under Review)**
In this PR, I created an OLMo Wrapper Class with regression, classification, and causal_lm tasks with 3 pytest files for classification, regression, and causal_lm. I used a linear layer via OlmoForSequenceClassification and loaded the model with float16, gradient checkpointing, and quantization for memory efficiency. In OlmoForSequenceClassification, I use a linear head over the last token's hidden state. Since OLMo is decoder-only, which means it doesn't inherently support regression and classification tasks, the head pools the hidden state at the last non-padding token of each sequence (via config.pad_token_id) instead of a [CLS] token. If labels are passed to forward, a loss is computed per config.problem_type. I use MSE loss for regression and BCEWithLogitsLoss for classification. Then in the Olmo class, I give the user the option to use either random weight initialization or to skip it by loading a pretrained model. I also give the user the options to use or don't use quantization and the same thing for gradient checkpointing. Quantization and gradient checkpointing are memory management techniques which are crucial to use a large OLMo model. Quantization reduces the numerical precision of the model’s weights to a lower numerical precision whereas gradient checkpointing trades a small amount of extra computation for a big reduction in GPU memory usage. I then had to override the load_from_pretrained() and prepare_batch() functions from hf_models.py because I was using multitask regression and multitask classification which isn't in the original load_from_pretrained() and prepare_batch() functions from hf_models.py. I added multiple tests such as save and reload and overfit and they all pass.

### **PR #5071  (OLMo Wrapper Part 2 , Under Review)**
In this PR, I added support for Low-Rank Adaptation (LoRA) , QLoRA (Quantized LoRA) and Distributed Data Parallel (DDP). LoRA and QLoRA are both parameter‑efficient fine‑tuning (PEFT) methods for large models such as OLMo, but they differ in how they reduce memory and compute costs. LoRA reduces the number of trainable adapters by inserting low-rank matrices into selected layers whereas QLoRA combines LoRA with quantization of the base model weights by loading the base model in low-bit precision and having LoRA adapters trained on top. In olmo.py I added LoRA, QLoRA, and full_finetune under the finetune strategy parameter so that the user is able to choose what finetune strategy they want to use. I also had to build a BitsandBytes config for QLoRA and apply PEFT for optionally applying k-bit training preparation and LoRA/QLoRA adapters to the model. I then tested DDP support via the lightning module in DeepChem. DDP is a distributed training technique that replicates the model across multiple GPUs while sharding the training data across processes. Since OLMo is a large model, DDP enables training to scale across multiple GPUs.

### **PR #5079    (OLMo Benchmarking Script , Draft)**
In this PR, I did benchmarking using allenai/OLMo-1B-hf across 9 datasets: Delaney, Tox21, BBBP, HIV, SIDER, CLINTOX, FREESOLV, LIPO, and BACE. I first would do pretraining using 300 samples from each dataset and doing pretraining on each dataset for 5 epochs with a learning rate of 1e-5 and using the finetuning strategy=qlora. I would do pretraining for each dataset and save their respective backbones separately. For example, after pretraining on delaney, I would save the delaney backbone and then load it for finetuning on only the delaney dataset. After pretraining, I would then do finetuning on each dataset with their respective backbones for 10 epochs with a learning rate of 3e-5 and finetuning strategy = qlora. Since I was doing preliminary benchmarking, I used splitter=”random”. The attached screenshot contains my preliminary benchmarking results. 
<img width="977" height="754" alt="Screenshot 2026-08-23 232421" src="https://github.com/user-attachments/assets/2205d53d-26b7-4b5f-91ba-8f7d8a254120" />

 According to my results, allenai/OLMo-1B-hf had a positive initial performance on the classification tasks but not so much on the regression tasks. However, the results are promising since I only used 300 samples. I don't have results for HIV and SIDER since I got OOM errors. This makes sense since they're large datasets. Then, for a more direct comparison, I benchmarked allenai/OLMo-1B-hf  against ChemBERTa3 using the scaffold splitter for both models. Unlike random splitting, scaffold splitting separates molecules according to their molecular scaffolds, providing a more challenging evaluation of generalization to structurally distinct compounds. The attached screenshot contains my initial results.  
<img width="995" height="854" alt="Screenshot 2026-08-23 231651" src="https://github.com/user-attachments/assets/6b09ef0c-6423-4941-87ad-28f19407fda4" />

 According to these results, using splitter ="scaffold" decreased the overall ROC-AUC for classification datasets and increased the RMS scores for the regression datasets. Under the scaffold-split evaluation, ChemBERTa3 outperformed OLMo-1B on most datasets, particularly on the regression tasks. However, allenai/OLMo-1B-hf approached ChemBERTa3 on ClinTox and achieved a 1.0 percentage-point higher ROC-AUC on BACE in this preliminary run.

### **Future Work**
After GSoC is over, I would like PR's #5001 , #5071, and #5079 merged into DeepChem and I'd like to expand benchmarking from allenai/OLMo-1B-hf  to allenai/OLMo-7B-hf and make the benchmarking process more scientifically rigorous by increasing the number of pretraining samples, finetuning epochs, etc. 

### **Acknowledgements**
I'd like to thank @riya-singh28 , @Harindhar10  , @sudo-rsingh , and @rbharath for supporting me throughout this journey since it was the first time I ever contributed to open source. It was difficult but fun at the same time and I learned a lot. I feel confident that I'll be able to contribute to other organizations outside of DeepChem. 

### **Links**
1. Proposal: https://www.overleaf.com/read/svfcjvtwspwr#cebd75
2. GitHub: https://github.com/priyadp1
3. Personal Website: https://website-delta-self-50.vercel.app/
